design/README.md

```diff
@@ -0,0 +1,73 @@
+# Code specification
+
+## Motivation
+
+Replace the current APFELcomb code with a simpler and faster tool to generate FK Tables for NNPDF fits.
+
+## APFELcomb design and issues
+
+### Workflow
+
+The apfelcomb project compiles a program which connects several tools and
+databases from NNPDF and external providers in order to generate FK Tables,
+where observables are stored into tables with a predefined input scale energy value for PDFs evaluation.
+
+The user/develop should include the custom implementation of the combination operation for each specific process category. All operations involving DGLAP, DIS and Fixed-target DY are performed at runtime. Operations involving the generation of hadronic observables with or without electroweak corrections are performed off-line and stored on git-lfs repositories.
+
+The mapping between the requested FK table configuration, in terms of theory and dataset choice, are controlled by the NNPDF theory sqlite database and by a custom sqlite database build inside apfelcomb.
+
+### Inputs
+
+The apfelcomb program takes as input the:
+- the theory database from NNPDF,
+- a custom database with specifications for each dataset in NNPDF,
+- CommonData files from NNPDF/buildmaster,
+- APPLgrid grids,
+- PineAPPL grids.
+
+### External tools
+
+The current apfelcomb implementation relies on third-party tools for:
+- DGLAP evolution kernels (APFEL),
+- DIS observables (APFEL),
+- Fixed-target DY (APFEL)
+- APPLgrids for hadronic processes (APPLgrid),
+- PineAPPL grids for processes with EW corrections (PineAPPL).
+
+### Problems to be avoided
+
+Here a short summary of practical problems we would like to avoid:
+- avoid different data input format, today we have to maintain parsers for APFEL (DIS/FTDY), APPLgrid and PineAPPL,
+- avoid long runtime operations that can be cached and stored in a server a priori, e.g.: DGLAP evolution kernels, DIS and FTDY grids,
+- reduce to minimum data input processing,
+- delegate the FK Table construction to a single code/function,
+- provide tools to cache results and upload to server
+- provide tools to optimize the quality of FK tables.
+
+## Proposal
+
+We propose to develop a code which performs the combination and provide tools to prepare, store and update its requirements.
+
+### Input
+
+The program should take as input:
+1. the theory number identification number, using the NNPDF theory database. Note that in opposite to apfelcomb we should take the number and not the full sqlite string.
+2. the dataset name following the NNPDF convention.
```

Comment:
Here we need to decide if `eko` and `yadism` additional configs will be provided as:
- an additional database
- an additional configuration file
- connected to the respective datasets
- hardcoded in `Pineko` once for all

Reply:
Yes, but note that this program/step is independent of yadism.
We need another section in this docs for the specifications of the "theory generators" in particular (yadism, mg5).
We may follow the development pattern that @cschwan is using for mg5, and extend it if needed.

Reply:
I'm not aware of the way @cschwan is managing mg5 inputs at the moment, but I agree on having a separate db for "theory generators", it will be a unique source of truth and it will make the FKtables fully reproducible.

Reply:
@scarrazza: This is not exactly on-topic, but I was thinking about a *universal runcard generator*: A problem that we still have is that the theory predictions are calculated using very different programs, and usually it's required to have a precise knowledge of all the intricacies of those programs. Especially for `mg5_aMC` the amount of extra script lines that I had to write is not small. At some point we might want to leverage SHERPA, as not all the processes can be run with `mg5_aMC`, and let's not forget yadism.

I think a worthwhile project would be to write some kind of 'universal runcard generator', where we bundle our knowledge of all of those programs, where we can write a **single** specification for the PineAPPL grids that we'd like to generate, with all the cuts and parameters, and where we ideally can choose the theory program if multiple programs can calculate the process.

I wonder whether this is worth the effort. This is of course yet another instance of 'every problem can be solved be another level of indirection', but in the long run I think it this universal runcard generator might be worth having; at some point we surely want to generate NNLO grids, and the variety of programs there is even larger than at NLO.

Reply:
Yes, maybe you're are right that we should think carefully if it's needed or we are just adding extra complications, but my personal feeling is that it would be worth, because this is really the last missing link in making the runs fully reproducible.

I was thinking that it may be worth to design the specs as an abstract [schema](https://json-schema.org/) and then make it "pluggable", in the sense that we support some basic management, but then the bindings for each program are bundled as separate plugins (even if maybe shipped within the same project).

Then should a PineAPPLgrid be bundled with it's own generating runcard?

Reply:
Regarding the *universal runcard generator*, @cschwan, I'm surprised this is not something experimentalists are asking Sherpa and Madgraph (and other generators) developers to agree upon. I mean, there was an agreement for the output files after all. Do you know whether there are any efforts in this direction?

Reply:
I don't think there's something like this. In the (update to the Binoth) Les Houches Accord there's a specification for matrix elements, but that's not really what I'm thinking of. I suppose experimentalist would be *very* happy if such a program would exist.

Reply:
Indeed. Going even more off-topic (sorry) as this is maybe decoupled from this project but it would be interesting to provide such a tool (and therefore define such a standard). A sort of runcard reader that projects can use as a dependency (instead of using whatever they use).
Beyond the usefulness for experimentalists and us, I remember that back when I was doing fixed-orders there were bugs and inconsistencies that traced back to using some different parameter between NNLOJET and MCFM or Sherpa/Madgraph. And it was very much not obvious to find what was wrong.*

It is a complicated problem (political and otherwise) but it could be fun.

* going slightly on-topic, errors due to small inconsistencies did propagate to published NNLO results so it was very much non trivial until the end.

Reply:
I think it's actually not that difficult; what we have to do is to read runcards (our own format) and then write runcards and patches for the Monte Carlos (which is completely independent from what the MCs do, so no political problem) and start them.

As a starting point we could try to do that with Madgraph5 for which the `run.sh` script in https://github.com/scarrazza/applgridew basically already does it. The first step would be to port it to Python or Rust, and then to change the syntax of the runcards. For the time being we're only interested in PineAPPL grids, so there's also a well-defined output.

Reply:
Can't access that repo.
I mean it would be complicated to convince people to adopt the `pineko` standard. I was being ambitious there.

Reply:
@scarrazza : Can you give @scarlehoff access to it?

Reply:
I think you are referring to https://github.com/scarrazza/applgridphoton, but there is no `run.sh`, what were you pointing to?

Let me propose again to use json schema for the specs, there are bindings for both rust and python available.

Reply:
@AleCandido : No, it's https://github.com/scarrazza/applgridew, where we have all the patches and runcards for the generation of the PineAPPL grids. Something like JSON for our runcards is definitely a good idea, but that's already step 2 ;).

Reply:
Then I cannot access that repo...

Reply:
Now you (all) should have access to repo. I think that this universal generator is a great idea, and in fact the script from @cschwan is already a good starting point.